File: lib/exercise.rb
require 'date'
#require 'random'
require './lib/parts'

class Exercise
  attr_reader :index, :parts
  START_DATE = Date.new(2020, 11, 30)

  def initialize(index)
    @index = index
    populate
  end

  def self.by_date(date)
    index = (date - START_DATE).to_i
    self.new index
  end

  def date
    START_DATE + index
  end

  private

  def populate
    rnd = Random.new(@index)
    @parts = [
      MajorScale.from_number(rnd.rand(15)),
      MinorNaturalScale.from_number(rnd.rand(12)),
      MinorHarmonicScale.from_number(rnd.rand(12)),
      MajorBrokenChord.from_number(rnd.rand(12)),
      MinorBrokenChord.from_number(rnd.rand(12)),
    ]
  end
end

File: src/exercise.rs
use lib::*;
use super::parts::*;

use chrono::prelude::*;
use chrono::Duration;

pub struct Exercise {
  pub index: i32,
  pub parts: [Box<dyn Part>; 5]
}

impl Exercise {
  pub fn new(n: i32) -> Exercise {
    Exercise{
      index: n,
      parts: [
        Box::new(MajorScale::new(n)),
        Box::new(MinorNaturalScale::new(n)),
        Box::new(MinorHarmonicScale::new(n)),
        Box::new(MajorBrokenChord::new(n)),
        Box::new(MinorBrokenChord::new(n))
      ]
    }   
  }

  pub fn date(&self) -> Date<Utc> {
    Utc.ymd(2020, 11,30) + Duration::days(self.index as i64)
  }
}
